tests/MongoDB.Driver.Core.Tests/Core/Operations/AsyncCursorTests.cs

```diff
@@ -313,143 +311,151 @@ private ConnectionDescription CreateConnectionDescriptionSupportingSession()
+        public void Session_reference_count_should_be_decremented_as_soon_as_possible(int collectionSize, int batchSize)
         {
-            // private fields
-            private AsyncCursor<BsonDocument> _instance;
+            RequireServer.Check();
+            DropCollection();
+            var documents = Enumerable.Range(1, collectionSize).Select(n => new BsonDocument("_id", n));
+            Insert(documents);
 
-            // constructors
-            public Reflector(AsyncCursor<BsonDocument> instance)
+            var findOperation = new FindOperation<BsonDocument>(_collectionNamespace, BsonDocumentSerializer.Instance, _messageEncoderSettings)
             {
-                _instance = instance;
-            }
+                BatchSize = batchSize
+            };
 
-            // public properties
-            public int? BatchSize
+            using (var binding = new WritableServerBinding(CoreTestConfiguration.Cluster, _session.Fork()))
+            using (var cursor = findOperation.Execute(binding, CancellationToken.None))
             {
-                get
+                AssertExpectedChannelSourceDisposedValue(cursor);
+                while (cursor.MoveNext())
                 {
-                    var fieldInfo = _instance.GetType().GetField("_batchSize", BindingFlags.NonPublic | BindingFlags.Instance);
-                    return (int?)fieldInfo.GetValue(_instance);
+                    AssertExpectedChannelSourceDisposedValue(cursor);
                 }
+                AssertExpectedChannelSourceDisposedValue(cursor);
             }
+        }
 
-            public IChannelSource ChannelSource
-            {
-                get
-                {
-                    var fieldInfo = _instance.GetType().GetField("_channelSource", BindingFlags.NonPublic | BindingFlags.Instance);
-                    return (IChannelSource)fieldInfo.GetValue(_instance);
-                }
-            }
+        // private methods
+        private void AssertExpectedChannelSourceDisposedValue(IAsyncCursor<BsonDocument> cursor)
+        {
+            var cursorImplementation = (AsyncCursor<BsonDocument>)cursor;
+            var channelSource = cursorImplementation._channelSource();
+            var cursorId = cursorImplementation._cursorId();
+            var actualDisposed = GetDisposed(channelSource);
+            var expectedDisposed = cursorId == 0;
+            actualDisposed.Should().Be(expectedDisposed);
+        }
 
-            public CollectionNamespace CollectionNamespace
+        private bool GetDisposed(IChannelSource channelSource)
+        {
+            if (channelSource == null)
             {
-                get
-                {
-                    var fieldInfo = _instance.GetType().GetField("_collectionNamespace", BindingFlags.NonPublic | BindingFlags.Instance);
-                    return (CollectionNamespace)fieldInfo.GetValue(_instance);
-                }
+                return true;
```

Comment:
If it's null that means we've already disposed it, and because it's null we can't reflect on it below.